libtiledbsoma/src/soma/column_buffer.h

```diff
@@ -343,6 +251,238 @@ class ColumnBuffer {
+
+    CArrayColumnBuffer() = delete;
+    CArrayColumnBuffer(const CArrayColumnBuffer&) = delete;
+    CArrayColumnBuffer(CArrayColumnBuffer&&) = default;
+
+    ~CArrayColumnBuffer();
+
+    std::span<std::byte> data() override;
+    std::span<uint64_t> offsets() override;
+    std::span<uint8_t> validity() override;
+
+    std::span<const std::byte> data() const override;
+    std::span<const uint64_t> offsets() const override;
+    std::span<const uint8_t> validity() const override;
+
+   private:
+    std::unique_ptr<std::byte[]> data_;
+    std::unique_ptr<uint64_t[]> offsets_;
+    std::unique_ptr<uint8_t[]> validity_;
+};
+
+/**
+ * @brief Class to store data for a TileDB dimension or attribute.
+ *
+ */
+class VectorColumnBuffer : public ReadColumnBuffer {
+    // This "medium size" -- 1 GiB -- is a good balance between improved remote
+    // I/O performance (bigger = better) and friendliness for smaller hardware
+    // (e.g.  tiny CI runners). This value is good for general in-between
+    // hardware including modern laptops and a broad range of EC2 instances. CI
+    // can ask for smaller; power-server users can ask for larger.
+    inline static const size_t DEFAULT_ALLOC_BYTES = 1 << 30;
+    inline static const std::string CONFIG_KEY_INIT_BYTES = "soma.init_buffer_bytes";
+
+   public:
+    //===================================================================
+    //= public static
+    //===================================================================
+
+    /**
+     * @brief Create a ColumnBuffer from an array and column name.
+     *
+     * @param array TileDB array
+     * @param name TileDB dimension or attribute name
+     * @return ColumnBuffer
+     */
+    static std::shared_ptr<ColumnBuffer> create(std::shared_ptr<Array> array, std::string_view name);
+
+    //===================================================================
+    //= public non-static
+    //===================================================================
+
+    /**
+     * @brief Construct a new ColumnBuffer object
+     *
+     * @param name Column name
+     * @param type TileDB datatype
+     * @param num_cells Number of cells to allocate for offsets and validity
+     * @param num_bytes Number of bytes to allocate for data
+     * @param is_var Column type is variable length
+     * @param is_nullable Column can contain null values
+     * @param enumeration Optional Enumeration associated with column
+     * @param is_ordered Optional Enumeration is ordered
+     */
+    VectorColumnBuffer(
+        std::string_view name,
+        tiledb_datatype_t type,
+        size_t num_cells,
+        size_t num_bytes,
+        bool is_var = false,
+        bool is_nullable = false,
+        std::optional<Enumeration> enumeration = std::nullopt,
+        bool is_ordered = false);
+
+    VectorColumnBuffer() = delete;
+    VectorColumnBuffer(const VectorColumnBuffer&) = delete;
+    VectorColumnBuffer(VectorColumnBuffer&&) = default;
```

Comment:
I get the following warning when compiling:
```
/home/jules/Software/TileDB-Inc/TileDB-SOMA/libtiledbsoma/src/soma/column_buffer.h:474:5: warning: explicitly defaulted move constructor is implicitly deleted [-Wdefaulted-function-deleted]
  474 |     VectorColumnBuffer(VectorColumnBuffer&&) = default;
      |     ^
/home/jules/Software/TileDB-Inc/TileDB-SOMA/libtiledbsoma/src/soma/column_buffer.h:423:28: note: move constructor of 'VectorColumnBuffer' is implicitly deleted because base class 'ReadColumnBuffer' has a deleted move constructor
```